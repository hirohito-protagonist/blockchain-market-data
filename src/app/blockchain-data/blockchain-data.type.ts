import {
  fromBlockchainDataMarketPriceEntity,
  fromBlockchainDataStatisticsEntity
} from './entities/index.entity';

export enum DataServiceType {
  Ticker = 'ticker',
  ToBTC = 'tobtc',
  Stats = 'stats'
}

export interface BTCQuery {
  value: number | string;
  currency: string;
}

export type QueryType = BTCQuery | null;

export type DataResponseType = fromBlockchainDataMarketPriceEntity.MarketPricesEntities |
fromBlockchainDataStatisticsEntity.StatisticEntity | number;
