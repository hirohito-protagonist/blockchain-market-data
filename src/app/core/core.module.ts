import { NgModule, ModuleWithProviders } from '@angular/core';
import { StoreModule } from '@ngrx/store';

import { CoreAppContainerComponent } from './containers/app.container';

export {
  CoreAppContainerComponent
};

import { reducers } from './reducers/index.reducer';

@NgModule({
  imports: [
    StoreModule.forFeature('core', reducers)
  ],
  declarations: [ CoreAppContainerComponent ]
})
export class CoreModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: CoreModule,
      providers: []
    };
  }
}
