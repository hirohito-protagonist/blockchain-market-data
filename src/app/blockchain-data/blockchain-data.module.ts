import { NgModule, ModuleWithProviders } from '@angular/core';

import { StoreModule } from '@ngrx/store';

import { reducers } from './reducers/index.reducer';

@NgModule({
  imports: [
    StoreModule.forFeature('blockchain-data', reducers)
  ]
})
export class BlockchainDataModule {

  static forRoot(): ModuleWithProviders {

    return {
      ngModule: BlockchainDataModule,
      providers: []
    };
  }
}
