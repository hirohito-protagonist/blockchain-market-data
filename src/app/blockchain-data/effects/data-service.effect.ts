import { Injectable } from '@angular/core';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { fromDataServiceAction, fromBlockchainDataAction } from './../actions/index.action';
import { map, filter, switchMap, catchError } from 'rxjs/operators';
import { of, Observable } from 'rxjs';
import { DataServiceType, DataResponseType, ChartsQuery, BTCQuery } from './../blockchain-data.type';
import { ExchangeRatesService } from './../services/exchange-rates.service';
import { StatisticsService } from './../services/statistics.service';

@Injectable()
export class DataServiceEffects {

  @Effect()
  requestData$: Observable<Action> = this.action$.pipe(
    ofType(fromBlockchainDataAction.ActionTypes.FetchData),
    map((action: fromBlockchainDataAction.FetchData) => new fromDataServiceAction.Request({
      key: action.payload.key,
      query: action.payload.query
    }))
  );

  @Effect()
  responseToBTCData$: Observable<Action> = this.responseData$(
    DataServiceType.ToBTC,
    (action: fromBlockchainDataAction.FetchData) => this.exchangeRatesService.tobtc(
      (action.payload.query as BTCQuery).currency,
      (action.payload.query as BTCQuery).value as number
    ).pipe(
      map((response: number) => [action, response]),
      catchError(() => of([action, 0]))
    )
  );

  @Effect()
  responseTcikerData$: Observable<Action> = this.responseData$(
    DataServiceType.Ticker,
    (action: fromBlockchainDataAction.FetchData) =>
    this.exchangeRatesService.ticker().pipe(
      map((response) => [action, (response as DataResponseType)]),
      catchError(() => of([action, 0]))
    )
  );

  @Effect()
  responseStatsData$: Observable<Action> = this.responseData$(
    DataServiceType.Stats,
    (action: fromBlockchainDataAction.FetchData) =>
    this.statisticsService.stats().pipe(
      map((response) => [action, (response as DataResponseType)]),
      catchError(() => of([action, 0]))
    )
  );

  @Effect()
  responseChartsData$: Observable<Action> = this.responseData$(
    DataServiceType.Charts,
    (action: fromBlockchainDataAction.FetchData) =>
    this.statisticsService.charts(action.payload.query as ChartsQuery).pipe(
      map((response) => [action, (response as DataResponseType)]),
      catchError(() => of([action, 0]))
    )
  );

  constructor(
    private action$: Actions,
    private exchangeRatesService: ExchangeRatesService,
    private statisticsService: StatisticsService
  ) {}

  responseData$(serviceType: DataServiceType, fn: Function): Observable<Action> {
    return this.action$.pipe(
      ofType(fromBlockchainDataAction.ActionTypes.FetchData),
      filter((action: fromBlockchainDataAction.FetchData) => action.payload.key === serviceType),
      switchMap((action: fromBlockchainDataAction.FetchData) => fn(action)),
      map(([action, response]) => {
        return new fromDataServiceAction.Response({
          key: (action as fromBlockchainDataAction.FetchData).payload.key,
          response: (response as DataResponseType)
        });
      })
    );
  }
}
