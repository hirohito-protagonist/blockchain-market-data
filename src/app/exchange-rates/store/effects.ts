import { Injectable } from '@angular/core';
import { Actions, ofType, createEffect } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { convert, fetchData } from './../store/actions';
import { map } from 'rxjs/operators';
import { Observable } from 'rxjs';
import { fromBlockchainDataAction, DataServiceType } from '@bmd/blockchain-data';

@Injectable()
export class Effects {


  convertToBTC$: Observable<Action> = createEffect(() => this.action$.pipe(
    ofType(convert),
    map((action) => fromBlockchainDataAction.fetchData({
      key: DataServiceType.ToBTC,
      query: action.convert
    }))
  ));

  requestData$: Observable<Action> = createEffect(() => this.action$.pipe(
    ofType(fetchData),
    map(() => fromBlockchainDataAction.fetchData({
      key: DataServiceType.Ticker,
      query: null
    }))
  ));

  constructor(private action$: Actions) {}
}
