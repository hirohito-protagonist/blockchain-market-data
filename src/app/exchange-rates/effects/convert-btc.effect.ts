import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { fromConvertBtcAction } from './../actions/index.action';
import { map, switchMap, tap, catchError } from 'rxjs/operators';
import { of } from 'rxjs/observable/of';
import { ExchangeRatesService } from './../../shared/blockchain-api/blockchain-api.module';
import { fromBlockchainDataAction, DataServiceType } from './../../blockchain-data/blockchain-data.module';

@Injectable()
export class ConvertBtcEffects {

  @Effect()
  convertToBTC$: Observable<Action> = this.action$.pipe(
    ofType(fromConvertBtcAction.ActionTypes.Convert),
    map((action: fromConvertBtcAction.Convert) => new fromBlockchainDataAction.FetchData({
      key: DataServiceType.ToBTC,
      query: action.payload
    }))
  );

  constructor(private action$: Actions, private exchangeRatesService: ExchangeRatesService) {}
}
