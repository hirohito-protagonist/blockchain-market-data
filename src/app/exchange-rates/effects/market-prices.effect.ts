import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Effect, Actions, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';
import { fromMarketPricesAction } from './../actions/index.action';
import { map, switchMap, tap } from 'rxjs/operators';
import { fromBlockchainDataAction, DataServiceType } from '@bmd/blockchain-data';

@Injectable()
export class MarketPricesEffects {

  @Effect()
  requestData$: Observable<Action> = this.action$.pipe(
    ofType(fromMarketPricesAction.ActionTypes.FetchData),
    map(() => new fromBlockchainDataAction.FetchData({
      key: DataServiceType.Ticker,
      query: null
    }))
  );

  constructor(private action$: Actions) {}
}
