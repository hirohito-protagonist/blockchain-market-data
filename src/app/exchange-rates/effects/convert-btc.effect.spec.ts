import { HttpClientModule } from '@angular/common/http';
import { Actions } from '@ngrx/effects';
import { TestBed } from '@angular/core/testing';
import { cold, hot } from 'jasmine-marbles';

import { empty } from 'rxjs/observable/empty';
import { Observable } from 'rxjs/Observable';

import { ConvertBtcEffects } from './convert-btc.effect';
import { fromConvertBtcAction } from './../actions/index.action';
import { fromBlockchainDataAction, DataServiceType } from './../../blockchain-data/blockchain-data.module';

export class TestActions extends Actions {

  constructor() {
    super(empty());
  }

  set stream(source: Observable<any>) {
    this.source = source;
  }
}



export function getActions() {
  return new TestActions();
}


describe('ConvertBtcEffects', () => {

  let effects: ConvertBtcEffects;
  let actions$: TestActions;

  beforeEach(() => {

    TestBed.configureTestingModule({
      imports: [ HttpClientModule ],
      providers: [
        ConvertBtcEffects,
        { provide: Actions, useFactory: getActions }
      ]
    });

    effects = TestBed.get(ConvertBtcEffects);
    actions$ = TestBed.get(Actions);
  });

  it('should dispatch action to fetch data on convert action', () => {

    const action = new fromConvertBtcAction.Convert({ currency: 'USD', value: 200 });
    const completion = new fromBlockchainDataAction.FetchData({
      key: DataServiceType.ToBTC,
      query: {
        currency: 'USD',
        value: 200
      }
    });

    actions$.stream = hot('-a', { a: action });
    const expected = cold('-b', { b: completion });

    expect(effects.convertToBTC$).toBeObservable(expected);
  });
});
