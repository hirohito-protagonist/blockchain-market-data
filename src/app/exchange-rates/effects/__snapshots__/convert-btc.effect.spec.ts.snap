// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConvertBtcEffects should dispatch action to fetch data on convert action 1`] = `
Actions {
  "_isScalar": false,
  "operator": MapOperator {
    "project": [Function],
    "thisArg": undefined,
  },
  "source": Actions {
    "_isScalar": false,
    "operator": FilterOperator {
      "predicate": [Function],
      "thisArg": undefined,
    },
    "source": TestActions {
      "_isScalar": false,
      "source": HotObservable {
        "_isScalar": false,
        "error": undefined,
        "marbles": "-a",
        "source": SubscriptionLoggable {
          "_isScalar": false,
          "closed": false,
          "hasError": false,
          "isStopped": false,
          "messages": Array [
            Object {
              "frame": 10,
              "notification": Notification {
                "error": undefined,
                "hasValue": true,
                "kind": "N",
                "value": Convert {
                  "payload": Object {
                    "currency": "USD",
                    "value": 200,
                  },
                  "type": "[Exchange Rates Feature] Convert to BTC",
                },
              },
            },
          ],
          "observers": Array [],
          "scheduler": TestScheduler {
            "SchedulerAction": [Function],
            "actions": Array [],
            "active": false,
            "assertDeepEqual": [Function],
            "coldObservables": Array [
              SubscriptionLoggable {
                "_isScalar": false,
                "_subscribe": [Function],
                "messages": Array [
                  Object {
                    "frame": 10,
                    "notification": Notification {
                      "error": undefined,
                      "hasValue": true,
                      "kind": "N",
                      "value": FetchData {
                        "payload": Object {
                          "key": "tobtc",
                          "query": Object {
                            "currency": "USD",
                            "value": 200,
                          },
                        },
                        "type": "[Blockchain Data Feature API] Fetch data",
                      },
                    },
                  },
                ],
                "scheduler": [Circular],
                "subscriptions": Array [],
              },
            ],
            "flushTests": Array [],
            "frame": 0,
            "hotObservables": Array [
              [Circular],
            ],
            "index": -1,
            "maxFrames": 750,
            "now": [Function],
            "scheduled": undefined,
          },
          "subscriptions": Array [],
          "thrownError": null,
        },
        "values": Object {
          "a": Convert {
            "payload": Object {
              "currency": "USD",
              "value": 200,
            },
            "type": "[Exchange Rates Feature] Convert to BTC",
          },
        },
      },
    },
  },
}
`;
